clang/lib/StaticAnalyzer/Core/RangeConstraintManager.cpp

```diff
@@ -2883,22 +2883,16 @@ const llvm::APSInt *RangeConstraintManager::getSymVal(ProgramStateRef St,
 
 const llvm::APSInt *RangeConstraintManager::getSymMinVal(ProgramStateRef St,
                                                          SymbolRef Sym) const {
-  // TODO: Use `getRange()` like in `getSymVal()`, but that would make some
-  // of the reports of `BitwiseShiftChecker` look awkward.
-  const RangeSet *T = getConstraint(St, Sym);
-  if (!T || T->isEmpty())
-    return nullptr;
-  return &T->getMinValue();
+  auto &MutableSelf = const_cast<RangeConstraintManager &>(*this);
+  RangeSet Range = MutableSelf.getRange(St, Sym);
+  return Range.isEmpty() ? nullptr : &Range.getMinValue();
 }
 
 const llvm::APSInt *RangeConstraintManager::getSymMaxVal(ProgramStateRef St,
                                                          SymbolRef Sym) const {
-  // TODO: Use `getRange()` like in `getSymVal()`, but that would make some
-  // of the reports of `BitwiseShiftChecker` look awkward.
-  const RangeSet *T = getConstraint(St, Sym);
-  if (!T || T->isEmpty())
-    return nullptr;
-  return &T->getMaxValue();
+  auto &MutableSelf = const_cast<RangeConstraintManager &>(*this);
+  RangeSet Range = MutableSelf.getRange(St, Sym);
```

Comment:
Now that I think about it, why is `getRange` not a `const` method?

As far as I see, it returns a `RangeSet` by value and its typical usage looks like
```cpp
RangeSet New = getRange(St, Sym);
New = F.deletePoint(New, Point);
                                 
return setRange(St, Sym, New);
```
and while I didn't review its (complex, visitor-based) definition, semantically I don't think that it should modify anything.

Reply:
I had the same question. `SymbolicRangeInferrer` needs access to the  private `RangeSet::Factory F` member of the `RangeConstraintManager` object. That Factor does internal caching of the RangeSets it hands out.

Reply:
I see... if this factory caching doesn't have any externally visible non-performance effect, then I think it would be a textbook example where `mutable` is a good choice (or at least less bad than repeated use of `const_cast`).

What do you think about marking that private factory (or perhaps one of its members) as `mutable`?

Obviously I can understand it if you don't want to spend time on this additional NFC change; and I think this change is acceptable even with these `const_cast`s.

Reply:
IDK why I haven't thought about that, but now you say that it makes sense :D
Let me see what can I do here.

Reply:
Given this patch touches a sensitive component, I'd land it without spreading the const-correctness. It turns out other member functions could be also marked with `const` once this Factory member is mutable.
I'll come back to this later separately.

Reply:
I agree that this is probably better as a separate NFC change.

Reply:
The cleanup will follow in https://github.com/llvm/llvm-project/pull/112891